solr/core/src/java/org/apache/solr/metrics/SolrCoreMetricManager.java

```diff
@@ -105,30 +112,35 @@ public void loadReporters() {
   }
 
   /**
-   * Make sure that metrics already collected that correspond to the old core name are carried over
-   * and will be used under the new core name. This method also reloads reporters so that they use
-   * the new core name.
+   * Re-register all metric producers associated with this core. This recreates the metric registry
+   * resetting its state
    */
-  // NOCOMMIT SOLR-17458: Update for core renaming
-  public void afterCoreRename() {
-    assert core.getCoreDescriptor().getCloudDescriptor() == null;
-    String oldRegistryName = solrMetricsContext.getRegistryName();
-    String oldLeaderRegistryName = leaderRegistryName;
-    String newRegistryName =
-        createRegistryName(cloudMode, collectionName, shardName, replicaName, core.getName());
-    leaderRegistryName = createLeaderRegistryName(cloudMode, collectionName, shardName);
-    if (oldRegistryName.equals(newRegistryName)) {
-      return;
-    }
-    // close old reporters
-    metricManager.closeReporters(oldRegistryName, solrMetricsContext.getTag());
-    if (oldLeaderRegistryName != null) {
-      metricManager.closeReporters(oldLeaderRegistryName, solrMetricsContext.getTag());
-    }
-    solrMetricsContext =
-        new SolrMetricsContext(metricManager, newRegistryName, solrMetricsContext.getTag());
-    // load reporters again, using the new core name
-    loadReporters();
+  public void reregisterCoreMetrics() {
+    this.solrMetricsContext =
+        new SolrMetricsContext(
+            metricManager,
+            createRegistryName(cloudMode, collectionName, shardName, replicaName, core.getName()),
+            solrMetricsContext.getTag());
+    metricManager.removeRegistry(solrMetricsContext.getRegistryName());
+    if (leaderRegistryName != null) metricManager.removeRegistry(leaderRegistryName);
+
+    var attributesBuilder =
+        Attributes.builder()
+            .put(CORE_ATTR, core.getName())
```

Comment:
my point about the duplication, for example, can be seen in the attributes construction.  This is duplicated over in `registerMetricProcducer`.  But it's not a lot of duplication.  It could be ameliorated with comments on both sides.